### DISCUSSION TITLE
[Proposal] consolidating @swc/cli entrypoint for features

### DISCUSSION THREAD

**Discussion by kwonoj** [swc-project/swc#3041] [RFC]
## Description

This proposal attempts to describe proposals for potential changes to `@swc/cli`'s entrypoint.

Currently, `@swc/cli` offers differently named entrypoint for individual features. `swc` is for the main transpiler, `spack` (to be renamed to `swcpack` in v2) is for the bundler.

Instead of adding each entrypoint name for the each features, I'd like to propose consolidate `swc` as single entrypoint for any cli command execution. Each features can be represented as subcommand instead. Think how `git` works with its subcommand.

### Illustrated example

All of subcommands are pseudo, subject to change and only being used for example illustration.

- top-level command
```
$> swc // same as swc --help
$> swc --help

... list of subcommands and its description
```
For easier migration, V2 will prints out deprecated warning and forward args to `swc compile`. V3 will actually deprecate code paths to let top-level invocation for the compilation.

- Invoke compiler

```
// Identical to existing swc/cli commandline flags
swc compile --help
swc compile ./file.js
swc compile ...
```

- Invoke bundler
```
swc bundle --help
swc bundle ./file.js
```

#### Pros
- Single entrypoint to cli, no need to have new names for each features
- Easily extendable subcommand when adapt new features

#### Cons
- Slightly more complex args handling logics, but I guess pushing forward #1589 fully (making cli in fully-rust, does not invoke node.js at all) would allow better args parser.

**Comment by kdy1** [comment]
I love this idea. We can add `minify` too.

**Comment by kwonoj** [comment]
I'll start some work on the rust side for non-breaking changes to support these. Can slowly migrate things over @swc/cli for v2 and beyond once we have it.
